### PR TITLE
add split and overlapped snapshot test, fix node test snapshot send bug. 

### PR DESCRIPTION
**Pull request by siddontang**
@ngaut @BusyJay @disksing @tiancaiamao

### CODE REVIEW

File: tests/raftstore/test_split_region.rs [hunk 1]
```diff
@@ -249,3 +250,56 @@ fn test_server_delay_split_region() {
...
     test_delay_split_region(&mut cluster);
 }
+
+
+fn test_split_overlap_snapshot<T: Simulator>(cluster: &mut Cluster<T>) {
+    // We use three nodes 1, 2, 3 for this test.
+    cluster.bootstrap_region().expect("");
+    cluster.start();
+
+    // guarantee 1 is leader
```
Comment (ngaut) [inline tests/raftstore/test_split_region.rs]: s/1/node 1/  ?

File: src/raftstore/store/peer.rs [hunk 2]
```diff
@@ -502,6 +502,7 @@ impl Peer {
                                        -> Result<()> {
         metric_incr!("raftstore.send_raft_message");
         let mut send_msg = RaftMessage::new();
+        let is_snapshot = msg.get_msg_type() == raftpb::MessageType::MsgSnapshot;
```
Comment (ngaut) [inline src/raftstore/store/peer.rs]: Move down

File: tests/raftstore/test_split_region.rs [hunk 3]
```diff
@@ -249,3 +250,56 @@ fn test_server_delay_split_region() {
...
+    cluster.start();
+
+    // guarantee node 1 is leader
+    cluster.transfer_leader(1, util::new_peer(1, 1));
+    cluster.must_put(b"k0", b"v0");
+    assert_eq!(cluster.leader_of_region(1), Some(util::new_peer(1, 1)));
+
+    let pd_client = cluster.pd_client.clone();
+
+    // isolate 3 for region 1.
```
Comment (ngaut) [inline tests/raftstore/test_split_region.rs]: s/3/node 3/

**Comment by ngaut** [comment]
Rest LGTM

File: tests/raftstore/node.rs [hunk 4]
```diff
@@ -43,10 +45,26 @@ impl ChannelTransport {
...
+            Some(h) => {
+                try!(h.rl().send_raft_msg(msg));
+                if is_snapshot {
+                    // should report snapshot finish.
+                    self.routers
+                        .get(&from_store)
+                        .unwrap()
+                        .rl()
+                        .report_snapshot(region_id, to_peer_id, SnapshotStatus::Finish)
+                        .unwrap();
```
Comment (disksing) [inline tests/raftstore/node.rs]: Why not put `report Finish` together with `report Failure`?

Comment (siddontang) [inline tests/raftstore/node.rs]: where Failure? here is Finish.

Comment (disksing) [inline tests/raftstore/node.rs]: Can we move this to `peer.rs`?

**Comment by BusyJay** [comment]
LGTM

File: tests/raftstore/node.rs (hunk 4 above)
Comment (siddontang) [inline tests/raftstore/node.rs]: No, in partition test, we will wrap the send and may not call this function. 
so `peer.rs` can only know failure, send returning ok in `peer.rs` may not mean snapshot finish.

Comment (disksing) [inline tests/raftstore/node.rs]: So if send returns ok in `peer.rs` and snapshot does not finish. We will not report Finish nor Failure?

Comment (disksing) [inline tests/raftstore/node.rs]: All right. LGTM.

**Pull request merged**